File: frontend/src/api/reports.ts
import { api } from '@/api/client';
import type { CategoryKind, CategoryRankingReport, CategorySpendingReport } from '@/types';

export function fetchCategorySpendingReport(categoryId: number, startDate?: string, endDate?: string) {
  const params = new URLSearchParams({ category_id: String(categoryId) });
  if (startDate) params.set('start_date', startDate);
  if (endDate) params.set('end_date', endDate);
  return api.get<CategorySpendingReport>(`/reports/category-spending?${params.toString()}`);
}

export function fetchCategoryRanking(kind: CategoryKind, startDate?: string, endDate?: string) {
  const params = new URLSearchParams({ kind });
  if (startDate) params.set('start_date', startDate);
  if (endDate) params.set('end_date', endDate);
  return api.get<CategoryRankingReport>(`/reports/category-ranking?${params.toString()}`);
}

File: backend/app/api/routes/reports.py
from datetime import date as date_

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_session
from app.models.enums import CategoryKind
from app.schemas.reports import CategoryRankingReport, CategorySpendingReport
from app.services.reports_service import get_category_ranking_report, get_category_spending_report

router = APIRouter(prefix='/reports', tags=['reports'])


@router.get('/category-spending', response_model=CategorySpendingReport)
async def read_category_spending_report(
  category_id: int,
  start_date: date_ | None = Query(default=None),
  end_date: date_ | None = Query(default=None),
  session: AsyncSession = Depends(get_session),
) -> CategorySpendingReport:
  return await get_category_spending_report(session, category_id, start_date, end_date)


@router.get('/category-ranking', response_model=CategoryRankingReport)
async def read_category_ranking_report(
  kind: CategoryKind = CategoryKind.EXPENSE,
  start_date: date_ | None = Query(default=None),
  end_date: date_ | None = Query(default=None),
  session: AsyncSession = Depends(get_session),
) -> CategoryRankingReport:
  return await get_category_ranking_report(session, kind, start_date, end_date)
